src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,314 @@
+        }
+
+        public void SendDuplexStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+
+            SendCore(client, message);
+
+            StreamReader = new HttpContextClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void Dispose()
+        {
+            if (!Disposed)
+            {
+                Disposed = true;
+
+                _callCts.Cancel();
+                _callCts.Dispose();
+                _httpResponse?.Dispose();
+                StreamReader?.Dispose();
+                ClientStreamWriter?.Dispose();
+            }
+        }
+
+        private void SendCore(System.Net.Http.HttpClient client, HttpRequestMessage message)
+        {
+            SendTask = client.SendAsync(message, HttpCompletionOption.ResponseHeadersRead, _callCts.Token);
+        }
+
+        private HttpContentClientStreamWriter<TRequest, TResponse> CreateWriter(HttpRequestMessage message)
+        {
+            TaskCompletionSource<Stream> writeStreamTcs = new TaskCompletionSource<Stream>(TaskCreationOptions.RunContinuationsAsynchronously);
+            TaskCompletionSource<bool> completeTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            message.Content = new PushStreamContent(
+                (stream) =>
+                {
+                    writeStreamTcs.SetResult(stream);
+                    return completeTcs.Task;
+                },
+                GrpcContentTypeHeader);
+
+            var writer = new HttpContentClientStreamWriter<TRequest, TResponse>(this, writeStreamTcs.Task, completeTcs);
+            return writer;
+        }
+
+        private const string GrpcDeadlineHeader = "grpc-timeout";
+
+        private HttpRequestMessage CreateHttpRequestMessage()
+        {
+            var message = new HttpRequestMessage(HttpMethod.Post, Method.FullName);
+            message.Version = new Version(2, 0);
+
+            if (Options.Headers != null && Options.Headers.Count > 0)
+            {
+                foreach (var entry in Options.Headers)
+                {
+                    if (entry.Key == GrpcDeadlineHeader)
+                    {
+                        continue;
+                    }
+
+                    var value = entry.IsBinary ? Convert.ToBase64String(entry.ValueBytes) : entry.Value;
+                    message.Headers.Add(entry.Key, value);
+                }
+            }
+
+            if (Options.Deadline != null && Options.Deadline != DateTime.MaxValue)
+            {
+                var deadline = Options.Deadline.Value - _clock.UtcNow;
+
+                // JamesNK(todo) - Replicate C core's logic for formatting grpc-timeout
+                message.Headers.Add(GrpcDeadlineHeader, Convert.ToInt64(deadline.TotalMilliseconds) + "m");
```

Comment:
@jtattermusch Could you point me towards where C core sets the grpc-timeout header? I want to replicate it here.

Reply:
https://github.com/grpc/grpc/blob/master/src/core/lib/transport/timeout_encoding.cc